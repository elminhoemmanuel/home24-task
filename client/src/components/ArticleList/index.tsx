import React, { useState } from 'react'
import { Category } from '../../utils/types/productList';
import ArticleCard from "../ArticleCard";
import Pagination from '../Pagination';

type ArticleListProps = {
    categories: Category[];
    toggleSidebar: () => void;
}

const ArticleList = ({ categories, toggleSidebar }: ArticleListProps) => {

    const incrFactor = 20;
    const totalArticles = 50;
    const [range, setRange] = useState(20);
    const viewMore = ():void =>setRange(prev=>prev + incrFactor);
    const viewLess = ():void =>setRange(prev=>prev - incrFactor);

    const renderArticles = (categories: Category[]) => {
        const articles = categories?.map((category) => {
            return category?.categoryArticles.articles?.filter((article, x)=> x < range).map((article) => {
                return <ArticleCard key={article.name} article={article} />;
            });
        });
        return articles
    }

    return (
        <div className='pl-6 md:pl-72 relative pt-3 pr-6 md:pr-8'>
            {categories?.length ? (
                <div className='flex items-center mb-5'>
                    <button onClick={() => toggleSidebar()} className='block md:hidden mr-3 text-black hover:text-hmred'>
                        <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor" strokeWidth={2}>
                            <path strokeLinecap="round" strokeLinejoin="round" d="M4 6h16M4 12h16M4 18h16" />
                        </svg>
                    </button>

                    <h1 data-testid='header' className='text-2xl font-bold'>
                        {categories[0]?.name}
                        <small data-testid='article-count'> ({categories[0]?.articleCount})</small>
                    </h1>
                </div>
            ) : (
                'Loading...'
            )}
            <div className='grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-7 mb-6'>{renderArticles(categories)}</div>
            
            { categories?.length > 0  && <Pagination totalArticles={totalArticles} range={range} viewLess={viewLess} viewMore={viewMore} /> }
        </div>
    )
}

export default ArticleList