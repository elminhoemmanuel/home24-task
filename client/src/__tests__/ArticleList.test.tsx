import React from 'react';
import { render, screen } from '@testing-library/react';
import { within } from '@testing-library/dom'
import ArticleList from '../components/ArticleList';

const categories = [
    {
        name: "Möbel",
        articleCount: 79279,
        categoryArticles: {
            articles: [
                {
                    images: [{ path: "https://cdn1.home24.net/images/media/catalog/product/200x200/png/m/a/matratzenbezug-smood-webstoff-180-x-200cm-3477221.webp" }],
                    name: "Premium Komfortmatratze Smood",
                    prices: { currency: "EUR", regular: { value: 54999 } },
                    variantName: "180 x 200cm"
                },
                {
                    images: [{ path: "https://cdn1.home24.net/images/media/catalog/product/200x200/png/m/a/matratzenbezug-smood-webstoff-180-x-200cm-3477221.webp" }],
                    name: "Premium Komfortmatratze Smood",
                    prices: { currency: "EUR", regular: { value: 54999 } },
                    variantName: "180 x 200cm"
                }
            ]
        },
        childrenCategories: [
            {
                name: "Wohnzimmer",
                urlPath: "kategorie/wohnzimmermoebel/"
            },
            {
                name: "Schlafzimmer",
                urlPath: "kategorie/schlafzimmermoebel/"
            }

        ]

    }
]


test("On render, check that ArticleList is rendered with correct header", () => {
    const toggleSidebar = ():void =>{}
    const { getByTestId } = render(<ArticleList toggleSidebar={toggleSidebar} categories={categories} />);
    const { getByText } = within(screen.getByTestId('header'))
    expect(getByText("Möbel")).toBeInTheDocument()

});

