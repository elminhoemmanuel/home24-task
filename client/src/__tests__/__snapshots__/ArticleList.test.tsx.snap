// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`On render, check that ArticleList list is rendered correctly 1`] = `
<div
  class="pl-6 md:pl-72 relative pt-3 pr-6 md:pr-8"
>
  <div
    class="flex items-center mb-5"
  >
    <button
      class="block md:hidden mr-3 text-black hover:text-hmred"
    >
      <svg
        class="h-6 w-6"
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 6h16M4 12h16M4 18h16"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <h1
      class="text-2xl font-bold"
    >
      Möbel
      <small>
         (
        79279
        )
      </small>
    </h1>
  </div>
  <div
    class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-7 mb-6"
  >
    <div
      class="rounded border border-hmash hover:border-hmred p-4 cursor-pointer text-sm"
    >
      <div
        class="mb-3"
      >
        <img
          alt="product image"
          class="h-52 md:h-48 w-full"
          src="https://cdn1.home24.net/images/media/catalog/product/200x200/png/m/a/matratzenbezug-smood-webstoff-180-x-200cm-3477221.webp"
        />
      </div>
      <p>
        Premium Komfortmatratze Smood
      </p>
      <p
        class="font-bold my-2"
      >
        549,99 €
      </p>
      <div
        class="mt-3 mb-2"
      >
        <button
          class="text-sm rounded text-white p-3 cursor-pointer bg-hmred hover:bg-orange-500 block w-full"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="rounded border border-hmash hover:border-hmred p-4 cursor-pointer text-sm"
    >
      <div
        class="mb-3"
      >
        <img
          alt="product image"
          class="h-52 md:h-48 w-full"
          src="https://cdn1.home24.net/images/media/catalog/product/200x200/png/m/a/matratzenbezug-smood-webstoff-180-x-200cm-3477221.webp"
        />
      </div>
      <p>
        Premium Komfortmatratze Smood
      </p>
      <p
        class="font-bold my-2"
      >
        549,99 €
      </p>
      <div
        class="mt-3 mb-2"
      >
        <button
          class="text-sm rounded text-white p-3 cursor-pointer bg-hmred hover:bg-orange-500 block w-full"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-center py-6"
  >
    <button
      class="block text-hmred"
    >
      See More
    </button>
  </div>
</div>
`;
